firedrake/assemble_expressions.py

```diff
@@ -1,645 +1,366 @@
+    spaces = tuple(c.function_space() for c in expr.coefficients)
+    if any(type(s.ufl_element()) is ufl.MixedElement for s in spaces if s is not None):
+        raise ValueError("Not expecting a mixed space at this point, "
+                         "did you forget to index a function with .sub(...)?")
+    if len(set(s.finat_element for s in spaces if s is not None)) != 1:
+        raise ValueError("All coefficients must be defined on the same space")
+    lvalue = expr.lvalue
+    rvalue = expr.rvalue
+    broadcast = isinstance(rvalue, ConstantValue)
+    broadcast |= isinstance(rvalue, firedrake.Constant) and rvalue.ufl_shape == ()
+    rshape = rvalue.ufl_shape or (1,)
+    if not broadcast and lvalue.dat.dim != rshape:
+        raise ValueError("Mismatching shapes between lvalue and rvalue in pointwise assignment")
+    rvalue, = map_expr_dags(LowerCompoundAlgebra(), [rvalue])
+    lvalue, rvalue = map_expr_dags(translator, [lvalue, rvalue])
+
+    indices = gem.indices(len(lvalue.shape))
+    lvalue = gem.Indexed(lvalue, indices)
+    if not broadcast:
+        rvalue = gem.Indexed(rvalue, indices)
+    if isinstance(expr, AugmentedAssign):
+        rvalue = {IAdd: gem.Sum(lvalue, rvalue),
+                  ISub: gem.Sum(lvalue, gem.Product(gem.Literal(-1), rvalue)),
+                  IMul: gem.Product(lvalue, rvalue),
+                  IDiv: gem.Division(lvalue, rvalue)}[type(expr)]
+    return preprocess_gem([lvalue, rvalue])
+
+
+def pointwise_expression_kernel(exprs, scalar_type):
+    """Compile a kernel for pointwise expressions.
+
+    :arg exprs: List of expressions, all on the same iteration set.
+    :arg scalar_type: Default scalar type (numpy.dtype).
+    :returns: a PyOP2 kernel for evaluation of the expressions."""
+    if len(set(e.lvalue.node_set for e in exprs)) > 1:
+        raise ValueError("All expressions must have same node layout.")
+    translator = Translator()
+    assignments = tuple(compile_to_gem(expr, translator) for expr in exprs)
+    prefix_ordering = tuple(OrderedDict.fromkeys(itertools.chain.from_iterable(
+        node.index_ordering()
+        for node in gem_traversal([v for v, _ in assignments])
+        if isinstance(node, gem.Indexed))))
+    impero_c = compile_gem(assignments, prefix_ordering=prefix_ordering,
+                           remove_zeros=False, allow_return_accumulate=False)
+    coefficients = translator.varmapping
+    args = []
+    for expr in exprs:
+        retval = coefficients.pop(expr.lvalue)
+        args.append(loopy.GlobalArg(retval.name, shape=retval.shape, dtype=expr.lvalue.dat.dtype))
+        for c in expr.rcoefficients:
+            if c.dat == expr.lvalue.dat:
+                continue
+            var = coefficients[c]
+            args.append(loopy.GlobalArg(var.name, shape=var.shape, dtype=c.dat.dtype))
+    knl = generate(impero_c, args, 16, scalar_type, kernel_name="expression_kernel")
```

Comment:
The precision argument in tsfc.loopys `generate` does not get used, I think. Maybe we can drop the argument?

Reply:
It is indirectly via `ctx.epsilon` in the translation of `gem.Constant`. I'd be happy to drop it though, or just to infer it from `numpy.finfo`.

Reply:
I refactored tsfc to just handle precision from scalar_type everywhere, so now this is gone.